File: android/app/src/test/java/app/chompass/ui/progress/WeightTrendTest.kt
package app.chompass.ui.progress

import app.chompass.parity.ParityFixtures
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset

class WeightTrendTest {
    @Test
    fun resolvePrefersLastViewedOverDefault() {
        assertEquals(TimeRange.MONTH, TimeRange.resolve("1M", "1Y"))
        assertEquals(TimeRange.YEAR, TimeRange.resolve(null, "1Y"))
        assertEquals(TimeRange.WEEK, TimeRange.resolve(null, null))
        assertEquals(TimeRange.WEEK, TimeRange.resolve("nope", "also-nope"))
        assertEquals(TimeRange.ALL_TIME, TimeRange.resolve(null, "All"))
    }

    @Test
    fun parityFixtureMatchesComputeWeightTrend() {
        val rootObj = ParityFixtures.readJson("weight-trend-expected.json")
        val windowDays = rootObj.getInt("windowDays")
        val minDays = rootObj.getInt("minDaysInWindow")
        val cases = rootObj.getJSONArray("cases")
        for (i in 0 until cases.length()) {
            val c = cases.getJSONObject(i)
            val inputsJson = c.getJSONArray("inputs")
            val inputs = (0 until inputsJson.length()).map { j ->
                val row = inputsJson.getJSONObject(j)
                WeightTrendInput(
                    at = Instant.parse(row.getString("date")),
                    weightKg = row.getDouble("weightKg"),
                )
            }
            val expectedJson = c.getJSONArray("expected")
            val expected = (0 until expectedJson.length()).map { j ->
                val row = expectedJson.getJSONObject(j)
                WeightTrendPoint(
                    day = LocalDate.parse(row.getString("date")),
                    valueKg = row.getDouble("valueKg"),
                )
            }
            val actual = computeWeightTrend(
                weighIns = inputs,
                zone = ZoneOffset.UTC,
                windowDays = windowDays,
                minDaysInWindow = minDays,
            )
            assertEquals("case ${c.getString("id")} size", expected.size, actual.size)
            for (k in expected.indices) {
                assertEquals("case ${c.getString("id")} day[$k]", expected[k].day, actual[k].day)
                assertEquals(
                    "case ${c.getString("id")} value[$k]",
                    expected[k].valueKg,
                    actual[k].valueKg,
                    1e-9,
                )
            }
        }
    }

    @Test
    fun splitTrendSegmentsBreaksLargeGaps() {
        val points = listOf(
            WeightTrendPoint(LocalDate.of(2026, 3, 2), 80.0),
            WeightTrendPoint(LocalDate.of(2026, 3, 3), 79.5),
            WeightTrendPoint(LocalDate.of(2026, 3, 20), 78.0),
            WeightTrendPoint(LocalDate.of(2026, 3, 21), 77.5),
        )
        val segments = splitTrendSegments(points, maxGapDays = 7)
        assertEquals(2, segments.size)
        assertEquals(2, segments[0].size)
        assertEquals(2, segments[1].size)
        assertTrue(segments[0].first().day == LocalDate.of(2026, 3, 2))
    }
}

File: web/app/src/lib/chompass-core/__tests__/weight-trend.test.js
// @ts-check
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import {
  computeWeightTrend,
  resolveProgressRangeId,
  splitTrendSegments,
} from "../weight-trend.js";

const here = dirname(fileURLToPath(import.meta.url));
const fixture = JSON.parse(
  readFileSync(join(here, "../../../../../../testdata/parity/weight-trend-expected.json"), "utf8"),
);

describe("weight trend (parity fixture)", () => {
  it("resolveProgressRangeId prefers last viewed", () => {
    assert.equal(resolveProgressRangeId("1M", "1Y"), "1M");
    assert.equal(resolveProgressRangeId(null, "1Y"), "1Y");
    assert.equal(resolveProgressRangeId(undefined, undefined), "1W");
    assert.equal(resolveProgressRangeId("nope", "also-nope"), "1W");
  });

  it("matches shared golden cases", () => {
    for (const c of fixture.cases) {
      const actual = computeWeightTrend(c.inputs, {
        windowDays: fixture.windowDays,
        minDaysInWindow: fixture.minDaysInWindow,
      });
      assert.equal(actual.length, c.expected.length, c.id);
      for (let i = 0; i < c.expected.length; i++) {
        assert.equal(actual[i].date, c.expected[i].date, `${c.id}[${i}].date`);
        assert.ok(
          Math.abs(actual[i].valueKg - c.expected[i].valueKg) < 1e-9,
          `${c.id}[${i}].valueKg ${actual[i].valueKg} vs ${c.expected[i].valueKg}`,
        );
      }
    }
  });

  it("splitTrendSegments breaks large gaps", () => {
    const segments = splitTrendSegments(
      [
        { date: "2026-03-02", valueKg: 80 },
        { date: "2026-03-03", valueKg: 79.5 },
        { date: "2026-03-20", valueKg: 78 },
        { date: "2026-03-21", valueKg: 77.5 },
      ],
      7,
    );
    assert.equal(segments.length, 2);
    assert.equal(segments[0].length, 2);
    assert.equal(segments[1].length, 2);
  });
});
